File: crates/xmtp_sdk/conformance/kotlin/src/main/kotlin/LiftCost.kt
import uniffi.xmtp_sdk.FfiConverterTypeMessage
import uniffi.xmtp_sdk.FfiConverterTypeMessageData
import uniffi.xmtp_sdk.MessageData

// Deferred performance gate. Do not call until the owner enables benchmarks.
fun lift_cost_within_5_percent(sample: MessageData) {
    val count = 10_000
    val runs = 20
    val plain = mutableListOf<Long>()
    val hosted = mutableListOf<Long>()
    var observed = 0
    repeat(runs) {
        val startPlain = System.nanoTime()
        repeat(count) {
            val record = FfiConverterTypeMessageData.lift(FfiConverterTypeMessageData.lower(sample))
            observed += record.id.toString().length
        }
        plain += System.nanoTime() - startPlain

        val startHost = System.nanoTime()
        repeat(count) {
            val message = FfiConverterTypeMessage.lift(FfiConverterTypeMessageData.lower(sample))
            observed += message.id.toString().length
        }
        hosted += System.nanoTime() - startHost
    }
    check(observed > 0)
    check(hosted.sorted()[runs / 2] <= plain.sorted()[runs / 2] * 1.05)
}

File: crates/xmtp_sdk/conformance/swift/Sources/Conformance/LiftCost.swift
import Foundation
@testable import XmtpSdk

/// Deferred performance gate. Do not call until the owner enables benchmarks.
func lift_cost_within_5_percent(_ sample: MessageData) throws {
    let count = 10000
    let runs = 20
    var bytes: [UInt8] = []
    FfiConverterTypeMessageData.write(sample, into: &bytes)
    let encoded = Data(bytes)
    var plain: [Double] = []
    var hosted: [Double] = []
    var observed = 0
    for _ in 0 ..< runs {
        let startPlain = Date.timeIntervalSinceReferenceDate
        for _ in 0 ..< count {
            var buffer = (data: encoded, offset: encoded.startIndex)
            let record = try FfiConverterTypeMessageData.read(from: &buffer)
            observed += record.id.description.count
        }
        plain.append(Date.timeIntervalSinceReferenceDate - startPlain)

        let startHost = Date.timeIntervalSinceReferenceDate
        for _ in 0 ..< count {
            var buffer = (data: encoded, offset: encoded.startIndex)
            let message = try FfiConverterTypeMessage.read(from: &buffer)
            observed += message.id.description.count
        }
        hosted.append(Date.timeIntervalSinceReferenceDate - startHost)
    }
    guard observed > 0 else { throw ConformanceFailure("message lift produced no values") }
    let medianPlain = plain.sorted()[runs / 2]
    let medianHost = hosted.sorted()[runs / 2]
    guard medianHost <= medianPlain * 1.05 else {
        throw ConformanceFailure("message lift exceeded the 5 percent budget")
    }
}
